### ISSUE TITLE
[QUESTION] Gradients don't seem to propagate through array of structs

### ISSUE THREAD

**Issue by charles-cyluo** [NVIDIA/warp#1174]
Hi,

I've created an example of a `wp.struct` consisting of a single scalar attribute.
I launch a kernel that copies the value from the scalar array into the struct array, and then calculate identity loss of the struct array, but the loss doesn't seem to be propagated in to the scalar array.
```python
import warp as wp
wp.config.verify_autograd_array_access = True

device = "cuda:0"

@wp.struct
class Scalar:
    a: wp.float32

@wp.kernel
def pack_kernel(x: wp.array(dtype=wp.float32), y: wp.array(dtype=Scalar)):
    i = wp.tid()
    y[i].a = x[i]

@wp.kernel
def loss_kernel(y: wp.array(dtype=Scalar), loss: wp.array(dtype=wp.float32)):
    i = wp.tid()
    loss[i] = y[i].a

wp.init()
x = wp.ones(1, dtype=wp.float32, device=device, requires_grad=True)
print("x:", x.numpy())
y = wp.zeros(1, dtype=Scalar, device=device, requires_grad=True)
loss = wp.zeros(1, dtype=wp.float32, device=device, requires_grad=True)
tape = wp.Tape()
with tape:
    wp.launch(kernel=pack_kernel, dim=1, inputs=[x], outputs=[y], device=device)
    print("y:", y.numpy())
    wp.launch(kernel=loss_kernel, dim=1, inputs=[y], outputs=[loss], device=device)
print("loss:", loss.numpy())
tape.backward(loss=loss)
print("y.grad:", y.grad.numpy())
print("x.grad:", x.grad.numpy())
```
I expect `x.grad` to be 1, but output reveals that it is 0 despite y.grad being 1:
```
Warp 1.1.0 initialized:
   CUDA Toolkit 11.5, Driver 13.0
   Devices:
     "cpu"      : "x86_64"
     "cuda:0"   : "NVIDIA GeForce RTX 4090" (24 GiB, sm_89, mempool enabled)
   Kernel cache:
     /home/xxx/.cache/warp/1.1.0
x: [1.]
Module __main__ load on device 'cuda:0' took 167.64 ms
y: [(1.,)]
loss: [1.]
y.grad: [(1.,)]
x.grad: [0.]
```

I'm wondering whether this is an expected behavior.

**Comment by daedalus5** [comment]
Hi @charles-cyluo , thanks for submitting this. This is a longstanding bug that needs to be addressed. In the meantime, could you switch to a struct-of-arrays (SoA) approach instead?

```python
import warp as wp

wp.config.verify_autograd_array_access = True
wp.init()

device = "cuda:0"

@wp.struct
class SoA:
    a: wp.array(dtype=wp.float32)

@wp.kernel
def pack_kernel_soa(x: wp.array(dtype=wp.float32), y: SoA):
    i = wp.tid()
    y.a[i] = x[i]

@wp.kernel
def loss_kernel_soa(y: SoA, loss: wp.array(dtype=wp.float32)):
    i = wp.tid()
    wp.atomic_add(loss, 0, y.a[i])

x = wp.ones(1, dtype=wp.float32, device=device, requires_grad=True)
print("x:", x.numpy())

y = SoA()
y.a = wp.zeros(1, dtype=wp.float32, device=device, requires_grad=True)
loss = wp.zeros(1, dtype=wp.float32, device=device, requires_grad=True)

tape = wp.Tape()
with tape:
    wp.launch(kernel=pack_kernel_soa, dim=1, inputs=[x], outputs=[y.a], device=device)
    wp.launch(kernel=loss_kernel_soa, dim=1, inputs=[y], outputs=[loss], device=device)

print("y.a:", y.a.numpy())
print("loss:", loss.numpy())

tape.backward(loss=loss)

print("y.a.grad:", y.a.grad.numpy())
print("x.grad:", x.grad.numpy())
```

**Comment by charles-cyluo** [comment]
Thanks for the suggestion. The SoA approach works well on my side as a workaround.

**Comment by daedalus5** [comment]
@charles-cyluo gradient propagation through an array of structs should be supported as of commit https://github.com/NVIDIA/warp/commit/560ad813bdb7a4d5db330cbfc09eb18b7631d691

Please let us know if you run into any issues.
